Fix inplace operation bug, add tests

jwa7 opened the pull request:
Fixes #206 #207 

<!-- readthedocs-preview equistore start -->
----
:books: Documentation preview :books:: https://equistore--208.org.readthedocs.build/en/208/

<!-- readthedocs-preview equistore end -->

Luthaf commented on python/tests/operations/add.py:
```diff
@@ -292,6 +300,37 @@ def test_self_add_error(self):
             str(cm.exception),
             "B should be a TensorMap or a scalar value. ",
         )
+        self.assertTrue(equistore.equal(A, A_copy))  # check not modified in place
```
I'm not sure we need to check every where that nothing have been modified, this is already checked in the two test functions above

Luthaf commented on python/tests/operations/multiply.py:
```diff
@@ -298,6 +310,42 @@ def test_self_multiply_error(self):
             str(cm.exception),
             "B should be a TensorMap or a scalar value. ",
         )
+        self.assertTrue(equistore.equal(A, A_copy))
```
same comment as for `add`, we don't need to check this in all tests

Luthaf commented on python/src/equistore/operations/multiply.py:
```diff
@@ -34,10 +34,11 @@ def multiply(A: TensorMap, B: Union[float, TensorMap]) -> TensorMap:
     :return: New :py:class:`TensorMap` with the same metadata as ``A``.
     """
-
+    A = A.copy()
```
I don't think this is necessary. The code below (_multiply_block_constant/_multiply_block_block) should not make any in-place modification, so this would just be wasting memory

Luthaf reviewed (changes requested):
I don't think unconditionally copying the input is the best way to fix the problem, since it introduces a additional large cost to these operations. 

Since only `substract` has the bug, we should fix substract by either writing it in a similar way as `add`, or as `add(A, mul(B, -1.0))`. My preference would be to write it in a similar way to add: there is a bit more code duplication but we control the performance characteristics.

jwa7 commented:
Thanks for the comments @Luthaf . I have removed the unnecessary copying of TensorMaps, removed the redundant test case in each of the files, and slimmed down the number of places I perform the `self.assertTrue(equistore.equal(A, A_copy))` checks.

Luthaf commented on python/src/equistore/operations/subtract.py:
```diff
@@ -30,15 +31,9 @@ def subtract(A: TensorMap, B: Union[float, TensorMap]) -> TensorMap:
-                B.block(key).gradient(parameter).data[:] = (
-                    -B.block(key).gradient(parameter).data[:]
-                )
+        B = multiply(A=B, B=-1)
```
```suggestion
        B = multiply(B, -1)
```

Luthaf replied:
So this is fine for now, but also not the best solution: we are still creating a new tensormap to hold `-B`, but we don't have to. I'll open an issue to improve later

jwa7 commented:
Once #210 has been merged, this branch should be rebased and updated to ensure the updated `pow` tests are there
